src/solvers/solver_hardness.cpp

```diff
@@ -0,0 +1,324 @@
+      // on the same line: compare columns
+      auto left_column = string2optional_size_t(id2string(left.get_column()));
+      auto right_column = string2optional_size_t(id2string(right.get_column()));
+      if(!left_column.has_value() || !right_column.has_value())
+        return false;
+      if(*left_column < *right_column)
+        return true;
+      if(*left_column > *right_column)
+        return false;
+
+      return false; //equal
+    });
+
+  for(const auto &stat : hardness_stats)
+  {
+    auto hardness_stats_json = json_objectt{};
+    hardness_stats_json["SSA_expr"] = json_stringt{stat.ssa_expression};
+    hardness_stats_json["GOTO"] =
+      json_stringt{goto_instruction2string(stat.pc, ns)};
+
+    // It might be desirable to collect all SAT hardness statistics pertaining
+    // to a particular GOTO instruction, since there may be a number of SSA
+    // steps per GOTO instruction. The following JSON contains a unique
+    // identifier for each one.
+    hardness_stats_json["GOTO_ID"] = json_numbert{
+      std::to_string(reinterpret_cast<std::uintptr_t>(&(*stat.pc)))};
+    hardness_stats_json["Source"] = json(stat.pc->source_location);
+
+    auto sat_hardness_json = json_objectt{};
+    sat_hardness_json["Clauses"] =
+      json_numbert{std::to_string(stat.sat_hardness.clauses)};
+    sat_hardness_json["Literals"] =
+      json_numbert{std::to_string(stat.sat_hardness.literals)};
+    sat_hardness_json["Variables"] =
+      json_numbert{std::to_string(stat.sat_hardness.variables.size())};
+
+    hardness_stats_json["SAT_hardness"] = sat_hardness_json;
+
+    json_stream_array.push_back(hardness_stats_json);
+  }
+
+  if(!assertion_stats.empty())
+  {
+    auto assertion_stats_json = json_objectt{};
+    assertion_stats_json["SSA_expr"] =
+      json_stringt{assertion_stats.ssa_expression};
+
+    auto assertion_stats_pcs_json = json_arrayt{};
+    for(const auto pc : assertion_stats.pcs)
+    {
+      auto assertion_stats_pc_json = json_objectt{};
+      assertion_stats_pc_json["GOTO"] =
+        json_stringt{goto_instruction2string(pc, ns)};
+      assertion_stats_pc_json["Source"] = json(pc->source_location);
+      assertion_stats_pcs_json.push_back(assertion_stats_pc_json);
+    }
+    assertion_stats_json["Sources"] = assertion_stats_pcs_json;
+
+    auto assertion_hardness_json = json_objectt{};
+    assertion_hardness_json["Clauses"] =
+      json_numbert{std::to_string(assertion_stats.sat_hardness.clauses)};
+    assertion_hardness_json["Literals"] =
+      json_numbert{std::to_string(assertion_stats.sat_hardness.literals)};
+    assertion_hardness_json["Variables"] = json_numbert{
+      std::to_string(assertion_stats.sat_hardness.variables.size())};
+
+    assertion_stats_json["SAT_hardness"] = assertion_hardness_json;
+
+    json_stream_array.push_back(assertion_stats_json);
+  }
+}
+
+std::string solver_hardnesst::goto_instruction2string(
+  goto_programt::const_targett pc,
+  const namespacet &ns) const
+{
```

Comment:
I'm not completely sold on the idea of repeating so much of the printing logic here. I get the existing printing functionality didn't quite meet our purposes here, but it still feels a bit "off" to repeat this ad-hoc over here hidden in a private function somewhere.